keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,516 @@
+  RestartReplicatedJob FailurePolicyAction = "RestartReplicatedJob"
+  
+  // The failed child job is permanently failed and should not be
+  // restarted, active child jobs will continue to run to completion.
+  FailJob FailurePolicyAction = "FailJob"
+
+  // Restart the failed child job only, not the whole JobSet.
+  RestartJob FailurePolicyAction = "RestartJob"
+
+)
+
+// FailurePolicyRule defines a FailurePolicyAction to be executed if a child job
+// fails due to a reason listed in OnJobFailureReasons.
+type FailurePolicyRule struct {
+  // The action to take if the rule is matched.
+  // +kubebuilder:validation:Enum:=FailJobSet;Ignore;FailJob;RestartJob
+  Action FailurePolicyAction `json:"action"`
+  // The requirement on the job failure reasons. The requirement
+  // is satisfied if at least one reason matches the list.
+  // Each unique failure reason can only be associated with one FailurePolicyRule.
+  // An empty list applies the rule to any job failure reason.
+  // +kubebuilder:validation:UniqueItems:true
+  OnJobFailureReasons []string `json:"onJobFailureReasons"`
+  // TargetReplicatedJobs are the names of the replicated jobs the operator will apply to.
+  // A null or empty list will apply to all replicatedJobs.
+  // +optional
+  // +listType=atomic
+  // +kubebuilder:validation:UniqueItems
+  TargetReplicatedJobs []string `json:"targetReplicatedJobs,omitempty"`
+}
+
+type FailurePolicy struct {
+  // MaxRestarts defines the limit on the number of JobSet restarts.
+  // A restart is achieved by recreating all active child jobs.
+  MaxRestarts int32 `json:"maxRestarts,omitempty"`
+  // List of failure policy rules for this JobSet.
+  // For a given Job failure, the rules will be evaluated in order,
+  // and only the first matching rule will be executed.
+  Rules []FailurePolicyRule `json:"rules,omitempty"`
+}
+
+```
+
+### Constraints
+
+- For a given Job failure, the rules will be evaluated in order, and only the first matching rule will be executed. This
+follows an established pattern used by PodFailurePolicy in the Job controller.
+
+- OnJobFailureReasons must be a valid Job failure reason as defined [here](https://github.com/kubernetes/kubernetes/blob/2d4100335e4c4ccc28f96fac78153f378212da4c/staging/src/k8s.io/api/batch/v1/types.go#L537-L554)
+in the batchv1 Job API. At the time of writing this KEP, these include:
+  - `PodFailurePolicy`
+  - `BackoffLimitExceeded`
+  - `DeadlineExceeded`
+  - `MaxFailedIndexesExceeded`
+  - `FailedIndexes`
+
+### Implmentation
+
+The core part of the implementation will be defining what specific mechanisms the JobSet controller uses to implement
+the behavior defined for each FailurePolicyAction type:
+
+1) `FailJobSet`: To fail the JobSet immediately without restarting, the controller updates the JobSet status to failed.
+
+2) `Ignore`: To "ignore" the failure (i.e., restart the JobSet without counting it against `MaxRestarts`), the controller
```

Comment:
We need to use another annotation to track which jobs we need to restart since now we will not be incrementing it always, while restart-attempt will be a counter against maxRestarts.

Reply:
To restart all child jobs without incrementing restart-attempt, can't we simply delete the child jobs without incrementing the restart-attempt annotation, and allow the reconciliation process to recreate them? 

If we use a separate annotation, the controller will still ultimately react to the change in that annotation by deleting all the child jobs without incrementing the restart-attempt annotation.

The separate annotation could be used to set a limit / ensure we don't restart forever, but then we are just building a parallel version of the existing maxRestarts logic.

What is the benefit of the separate annotation here?

Reply:
> To restart all child jobs without incrementing restart-attempt, can't we simply delete the child jobs without incrementing the restart-attempt annotation, and allow the reconciliation process to recreate them?

How would you know if a child job was recreated or not? Remember that the jobset controller may restart in the middle of the reconciliation loop.

More concretely, in the following pseudo code which mimics what the jobset controller does:
```
for j in childJobs {
  recreate(j)
}
```
if the controller restarts before going through the list, the next reconcile call should continue from where we left off, not iterate over all the child jobs again.

Reply:
This is not yet addressed.

Reply:
Yeah sorry I had to shift my focus to something high priority that came up in another workstream. Thanks for the explanation, I will do another pass through the comments here today.

Reply:
no worries, I thought it fell through the cracks

Reply:
Updated this